atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownConcludedTransactions.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.knowledge;
+
+import com.google.common.collect.ImmutableRangeSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.Sets;
+import com.palantir.common.concurrent.CoalescingSupplier;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.concurrent.atomic.AtomicReference;
+
+@SuppressWarnings("UnstableApiUsage") // RangeSet usage
+public final class DefaultKnownConcludedTransactions implements KnownConcludedTransactions {
+    private static final SafeLogger log = SafeLoggerFactory.get(DefaultKnownConcludedTransactions.class);
+    private static final int MAX_ATTEMPTS = 20;
+
+    private final KnownConcludedTransactionsStore knownConcludedTransactionsStore;
+
+    /**
+     * Concurrency: All updates go through {@link #ensureRangesCached(RangeSet)}} and perform CASes to atomically
+     * evolve the value here. Copy on write should be acceptable given these range-sets are not expected to be large.
+     */
+    private final AtomicReference<ImmutableRangeSet<Long>> cachedConcludedTimestamps;
+
+    private final CoalescingSupplier<Void> cacheUpdater = new CoalescingSupplier<>(() -> {
+        updateCacheFromRemote();
+        return null;
+    });
+
+    public DefaultKnownConcludedTransactions(KnownConcludedTransactionsStore knownConcludedTransactionsStore) {
+        this.knownConcludedTransactionsStore = knownConcludedTransactionsStore;
+        this.cachedConcludedTimestamps = new AtomicReference<>(ImmutableRangeSet.of());
+    }
+
+    @Override
+    public boolean isKnownConcluded(long startTimestamp, Consistency consistency) {
+        if (cachedConcludedTimestamps.get().contains(startTimestamp)) {
+            return true;
+        }
+        if (consistency == Consistency.REMOTE_READ) {
+            return performRemoteReadAndCheckConcluded(startTimestamp);
+        }
+        return false;
+    }
```

Comment:
The docs made me think that we could not trust a false. Why will we not want to do a remote remove in this case?

Reply:
So this is the case where the consistency is `LOCAL_READ`, and `cachedConcludedTimestamps` does not contain the timestamp in question. The idea is that if this returns false, then we want to check if the provided timestamp is used in the transactions table. What do you mean by a remote remove here?

Reply:
oh i meant `remote refresh`